gneiss/balances.py

```diff
@@ -90,51 +102,63 @@ def balance_basis(tree_node):
 
 
 def _count_matrix(treenode):
-    n_tips = 0
-    nodes = list(treenode.levelorder(include_self=True))
-    # fill in the Ordered dictionary. Note that the
-    # elements of this Ordered dictionary are
-    # dictionaries.
-    counts = OrderedDict()
-    columns = ['k', 'r', 'l', 't', 'tips']
-    for n in nodes:
-        if n not in counts:
-            counts[n] = {}
-        for c in columns:
-            counts[n][c] = 0
-
-    # fill in r and l.  This is done in reverse level order.
-    for n in nodes[::-1]:
+    node_count = 0
+    for n in treenode.postorder(include_self=True):
+        node_count += 1
         if n.is_tip():
-            counts[n]['tips'] = 1
-            n_tips += 1
-        elif len(n.children) == 2:
-            lchild = n.children[0]
-            rchild = n.children[1]
-            counts[n]['r'] = counts[rchild]['tips']
-            counts[n]['l'] = counts[lchild]['tips']
-            counts[n]['tips'] = counts[n]['r'] + counts[n]['l']
+            n._tip_count = 1
         else:
-            raise ValueError("Not a strictly bifurcating tree!")
-
-    # fill in k and t
-    for n in nodes:
-        if n.parent is None:
-            counts[n]['k'] = 0
-            counts[n]['t'] = 0
-            continue
-        elif n.is_tip():
+            try:
+                left, right = n.children
+            except:
+                raise ValueError("Not a strictly bifurcating tree!")
+            n._tip_count = left._tip_count + right._tip_count
+
+    # TODO: use recarray
+    # col 0 -> right counts
+    # col 1 -> left counts
+    # col 2 -> k
+    # col 3 -> t
+    r_idx = 0
+    l_idx = 1
+    k_idx = 2
+    t_idx = 3
+    counts = np.zeros((node_count, 4), dtype=int)
+
+    for i, n in enumerate(treenode.levelorder(include_self=True)):
+        if n.is_tip():
             continue
-        # left or right child
-        # left = 0, right = 1
-        child_idx = 'l' if n.parent.children[0] != n else 'r'
-        if child_idx == 'l':
-            counts[n]['t'] = counts[n.parent]['t'] + counts[n.parent]['l']
-            counts[n]['k'] = counts[n.parent]['k']
+
+        n._lo_idx = i
+        node_counts = counts[i]
+
+        node_counts[r_idx] = 1 if n.is_tip() else n.children[1]._tip_count
+        node_counts[l_idx] = 1 if n.is_tip() else n.children[0]._tip_count
+
+        if n.is_root():
+            k = 0
+            t = 0
         else:
-            counts[n]['k'] = counts[n.parent]['k'] + counts[n.parent]['r']
-            counts[n]['t'] = counts[n.parent]['t']
-    return counts, n_tips
+            parent_counts = counts[n.parent._lo_idx]
+            if n is n.parent.children[0]:
+                #t = parent_counts[t_idx] + parent_counts[l_idx]
+                #k = parent_counts[k_idx]
+
+                k = parent_counts[k_idx] + parent_counts[r_idx]
```

Comment:
Tests don't pass if flipped though hence the confusion
On Jul 10, 2016 11:31 PM, "Jamie Morton" <notifications@github.com> wrote:

> In gneiss/balances.py
> <https://github.com/biocore/gneiss/pull/8#discussion_r70207160>:
>
> >          else:
> > -            counts[n]['k'] = counts[n.parent]['k'] + counts[n.parent]['r']
> > -            counts[n]['t'] = counts[n.parent]['t']
> > -    return counts, n_tips
> > +            parent_counts = counts[n.parent._lo_idx]
> > +            if n is n.parent.children[0]:
> > +                #t = parent_counts[t_idx] + parent_counts[l_idx]
> > +                #k = parent_counts[k_idx]
> > +
> > +                k = parent_counts[k_idx] + parent_counts[r_idx]
>
> This looks a little backwards. Stepping through the previous example I
> gave before
>
>                                        /-a
>                               /5-------|
>                              |          \-b
>                     /3-------|
>                    |         |          /-f
>                    |          \6-------|
>           /2-------|                    \-d
>          |         |
>          |         |          /-h
> -1-------|          \4-------|
>          |                    \-j
>          |
>           \-l
>
> If we were trying to calculate k and t for balance 4, and these values
> were all properly calculated for balance 2, the following will happen.
>
> k = parent_counts[k_idx] + parent_counts[r_idx]
>   = 0 + 2 = 2
> t = parent_counts[t_idx]
>   = 0
>
> When we know that k=1 and t=4 for the balance 4. Does this make more
> sense?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/gneiss/pull/8/files/23ad7ae6d93099e1a6575efac2e7500ac10833dd#r70207160>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAc8svn1j0If58AuHKZqblj3EKQdorOUks5qUeMrgaJpZM4JI_gR>
> .
>